### PR TITLE
RUN: Use `FormattingService` on 212

### PR DESCRIPTION
**Pull request by mchernyavsky**

### CODE REVIEW

File: src/211/main/kotlin/org/rust/ide/formatter/RustfmtExternalFormatProcessor.kt [hunk 1]
```diff
@@ -66,17 +65,12 @@ abstract class RustfmtExternalFormatProcessorBase : ExternalFormatProcessor {
...
+    override fun activeForFile(source: PsiFile): Boolean = isActiveForFile(source)
 
     // Never used by the platform?
     override fun indent(source: PsiFile, lineStartOffset: Int): String? = null
 
     companion object {
-        // BACKCOMPAT: 2021.1
-        private val BUILD_212: BuildNumber = BuildNumber.fromString("212")!!
-        // TODO: implement `FormattingService` instead
         private val reformatInProcess: ThreadLocal<Boolean> = ThreadLocal.withInitial { false }
```
Comment (Undin) [inline src/211/main/kotlin/org/rust/ide/formatter/RustfmtExternalFormatProcessor.kt:74]: I expect that this hack with thread local property will be removed in this PR

Comment (mchernyavsky) [reply src/211/main/kotlin/org/rust/ide/formatter/RustfmtExternalFormatProcessor.kt]: Fixed.

File: src/212/main/kotlin/org/rust/ide/formatter/RustfmtFormattingService.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import org.rust.stdext.enumSetOf
+import java.util.*
+
+@Suppress("UnstableApiUsage")
+class RustfmtFormattingService : AsyncDocumentFormattingService() {
+
+    override fun getFeatures(): Set<FormattingService.Feature> = FEATURES
+
+    override fun canFormat(file: PsiFile): Boolean =
+        file is RsFile && file.project.rustSettings.useRustfmt
```
Comment (vlad20012) [inline src/212/main/kotlin/org/rust/ide/formatter/RustfmtFormattingService.kt:35]:
```suggestion
        file is RsFile && file.project.rustSettings.useRustfmt && getFormattingReason() == FormattingReason.ReformatCode
```

**Review by vlad20012 — APPROVED** [review]
I found one issue. If I check `Reformat code Before commit` in the commit dialog
image
... then `RustfmtFormattingService` is used sometimes, at least when I commit a new file and this file doesn't contain a line break at the end. Also, since `RustfmtFormattingService`
 is **Async**, it actually reformats the document *after* committing, i.e. new changes are appearing in the changelist after such commit. Let's relive `getFormattingReason()` and check that it is `FormattingReason.ReformatCode` in `RustfmtFormattingService.canFormat`. I've checked that it works

File: src/212/main/kotlin/org/rust/ide/formatter/RustfmtFormattingService.kt (hunk 2 above)
Comment (mchernyavsky) [reply src/212/main/kotlin/org/rust/ide/formatter/RustfmtFormattingService.kt]: Fixed.

**Comment by mchernyavsky** [comment]
bors r=vlad20012

**Pull request merged**